### PR TITLE
External catalog

### PR DESCRIPTION
**Pull request by Maar-io**
Create new Catalog contract (previously called Base) to implement many-to-many relation, as in evm implementation for [RMRKCatalog](https://github.com/rmrk-team/evm/blob/master/contracts/RMRK/catalog/RMRKCatalog.sol)

- [x] Use Catalog as external contract
- [x] Change how Catalog(Base) is handled in the Rmrk contract. Add Catalog on assets.
- [x] Create new example contract
- [x] Unit test
- [x] Integration test update
- [x] update CI to use integration test

> Note! Some of the unit tests are removed since the equippable module now uses cross contract calls and can't be tested with ink::test. They will be added back with the ink::e2e tests.

### CODE REVIEW

File: crates/catalog/src/catalog.rs [hunk 1]
```diff
@@ -36,43 +33,43 @@ use openbrush::{
...
+                && (!part.equippable.is_empty() || part.is_equippable_by_all)
             {
                 return Err(RmrkError::BadConfig.into())
             }
-            self.data::<BaseData>().parts.insert(part_id, &part);
-            self.data::<BaseData>().part_ids.push(part_id);
-            self.data::<BaseData>().next_part_id += 1;
+            self.data::<CatalogData>().parts.insert(part_id, &part);
+            self.data::<CatalogData>().part_ids.push(part_id);
+            self.data::<CatalogData>().next_part_id += 1;
```
Comment (boyswan) [inline crates/catalog/src/catalog.rs:72]: We should stick to checked operations

File: crates/multiasset/src/lib.rs [hunk 2]
```diff
@@ -69,6 +74,7 @@ where
     #[modifiers(only_role(CONTRIBUTOR))]
     fn add_asset_entry(
         &mut self,
+        catalog_address: Option<AccountId>,
```
Comment (boyswan) [inline crates/multiasset/src/lib.rs:77]: Should this be `Option`? What does it mean to provide `None` here?

File: crates/multiasset/src/lib.rs [hunk 3]
```diff
@@ -261,6 +271,14 @@ where
             .get(&token_id)
             .unwrap_or_default())
     }
+
+    /// Fetch asset's catalog
+    fn get_asset_catalog(&self, asset_id: AssetId) -> Option<AccountId> {
```
Comment (boyswan) [inline crates/multiasset/src/lib.rs:276]: I think we should be explicit: `get_asset_catalog_address`

File: crates/catalog/src/catalog.rs [hunk 4]
```diff
@@ -108,40 +105,40 @@ where
...
 
         Ok(())
     }
 
-    /// Sets the metadata URI for Base
+    /// Sets the metadata URI for Catalog
     #[modifiers(only_role(CONTRIBUTOR))]
-    default fn setup_base(&mut self, base_metadata: String) -> Result<()> {
-        self.data::<BaseData>().base_uri = base_metadata;
+    default fn setup_catalog(&mut self, catalog_metadata: String) -> Result<()> {
```
Comment (boyswan) [inline crates/catalog/src/catalog.rs:115]: I know this is from the `base` implementation, but this does not need to be a `Result`

**Review by boyswan — COMMENTED** [review]
Looking good! Few minor comments. I also much prefer having `catalog::lib` as the modules only, perhaps we can do this for the other crates

File: crates/multiasset/src/lib.rs (hunk 3 above)
Comment (Maar-io) [reply crates/multiasset/src/lib.rs]: changed

File: crates/catalog/src/catalog.rs (hunk 1 above)
Comment (Maar-io) [reply crates/catalog/src/catalog.rs]: fixed

**Review by boyswan — APPROVED** [review]

**Pull request merged**